terraform/modules/minio/main.tf

```diff
@@ -0,0 +1,45 @@
+resource "juju_application" "minio" {
+  # Coordinator requires s3
+  name  = var.minio_app
+  model = var.model_name
+  trust = true
+
+  charm {
+    name    = var.minio_app
+    channel = var.channel
+  }
+  units = 1
+
+  config = {
+    access-key = var.minio_user
+    secret-key = var.minio_password
+  }
+}
+
+resource "null_resource" "s3management" {
+  triggers = {
+    # model_name = var.model_name
+    always_run = timestamp()
+  }
+  provisioner "local-exec" {
+    environment = {
+      MINIO_USER     = var.minio_user
+      MINIO_PASSWORD = var.minio_password
+    }
+    command = <<-EOT
+      bash "${path.module}/scripts/s3management.sh" \
+        --model-name ${var.model_name} \
+        --minio-app ${var.minio_app} \
+        --mc-binary-url ${var.mc_binary_url} \
+        --minio-url "http://${var.minio_app}-0.${var.minio_app}-endpoints.${var.model_name}.svc.cluster.local:9000" \
+        --loki-bucket ${var.loki.bucket_name} \
+        --mimir-bucket ${var.mimir.bucket_name} \
+        --tempo-bucket ${var.tempo.bucket_name} \
+        --loki-integrator ${var.loki.s3_integrator_name} \
+        --mimir-integrator ${var.mimir.s3_integrator_name} \
+        --tempo-integrator ${var.tempo.s3_integrator_name}
+    EOT
+  }
+
+  depends_on = [var.mimir, var.loki, var.tempo]
```

Comment:
According to the [TF docs](https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on#processing-and-planning-consequences):
>You should use depends_on as a last resort because it can cause Terraform to create more conservative plans that replace more resources than necessary. For example, Terraform may treat more values as unknown “(known after apply)” because it is uncertain what changes will occur on the upstream object. This is especially likely when you use depends_on for modules.

>Instead of depends_on, we recommend using [expression references](https://developer.hashicorp.com/terraform/language/expressions/references) to imply dependencies when possible. Expression references let Terraform understand which value the reference derives from and avoid planning changes if that particular value hasn’t changed, even if other parts of the upstream object have planned changes.

You are already using expression references for each module in lines 35 to 40 so I think like 44 is not relevant. This would require testing though!

Reply:
You're right, it is not necessary.

Removing and refactoring the shell script.